lib/_base_viewmodels.dart

```diff
@@ -66,63 +82,209 @@ abstract class ReactiveViewModel extends BaseViewModel {
+    Map<String, dynamic> _result = Map<String, dynamic>();
+    _streamDataMap
+        .forEach((key, streamData) => _result[key] = streamData.hasError);
+    return _result;
+  }
+
+  void _initialiseData() {
+    _streamDataMap = Map<String, StreamData>();
+  }
+
+  void runStreams() {
+    _initialiseData();
+    notifyListeners();
+    for (var key in streamsMap.keys) {
+      _streamDataMap[key] = setupStream(streamsMap[key]);
+      notifyListeners();
+    }
+  }
+}
+
+abstract class StreamViewModel<T> extends _SingleDataSourceViewModel<T> {
+  Stream<T> get stream;
+  // Holds the data, dataReady, hasError, isBusy properties
+  StreamData streamData;
+  get data => streamData.data;
+  get dataReady => streamData.dataReady;
+  get hasError => streamData.hasError;
+  get isBusy => streamData.isBusy;
+
+  //TODO: Integrate read/write for onData, onCancel, onSubscribed, onError and transformData properties
+  //TODO: Figure out a more elegant way of integrating StreamData
+  //TODO: Get this to work with generic stream types again
+  //TODO: Possibly have a default setbusy for overwrittenn lifecycle events
+
+  Future initialise() async {
+    streamData = setupStream(stream);
+  }
+}
+
+class StreamData<T> extends _SingleDataSourceViewModel<T> {
   Stream<T> stream;
+  StreamData(this.stream);
   StreamSubscription _streamSubscription;
 
-  StreamData(this.stream) : assert(stream != null);
-
-  factory StreamData.listen(stream,
-      {onData, onSubscribed, onError, onCancel, transformData}) {
-    StreamData streamData = StreamData(stream);
-    streamData._streamSubscription = stream.listen(
+  void initialise() {
+    _streamSubscription = stream.listen(
       (incomingData) {
-        streamData._hasError = false;
-        streamData.notifyListeners();
-        var interceptedData = streamData.transformData(incomingData);
+        _hasError = false;
+        notifyListeners();
+
+        var interceptedData = transformData(incomingData);
+
         if (interceptedData != null) {
-          streamData._data = interceptedData;
+          _data = interceptedData;
         } else {
-          streamData._data = incomingData;
+          _data = incomingData;
         }
 
-        streamData.notifyListeners();
-        streamData.onData(streamData._data);
+        notifyListeners();
+        onData(_data);
       },
       onError: (error) {
-        streamData._hasError = true;
-        streamData.onError(error);
-        streamData.notifyListeners();
+        _hasError = true;
+        _data = null;
+        onError(error);
+        notifyListeners();
       },
     );
-    streamData.onSubscribed();
-    return streamData;
+
+    onSubscribed();
   }
 
   /// Called when the new data arrives
   ///
   /// notifyListeners is called before this so no need to call in here unless you're
   /// running additional logic and setting a separate value.
-  void onData(T data) {}
+  void onData(T data) {
+    setBusy(false);
```

Comment:
We don't want any automatic busy states for streams. If you have multiple streams and one fails the model will be set to not busy automatically which might not be what the dev is expecting. For now we can remove all setBusy calls